### DISCUSSION TITLE
[Feature Request]: support for next `headers` and `cookies` functions

### DISCUSSION THREAD

**Discussion by JamesManningR** [storybookjs/storybook#25218] [Ideas]
### Is your feature request related to a problem? Please describe.

May be a duplicate of #19523

Using the next `headers` or `cookies` functions leaves you these errors respectively:

`Invariant: headers() expects to have requestAsyncStorage, none available.`
`Invariant: cookies() expects to have requestAsyncStorage, none available.

### Describe the solution you'd like

I think it would be ideal if the next plugin handled the usage of these functions out of the box

### Describe alternatives you've considered

**For anyone looking on how to solve the next headers/cookies and looking for a quick fix: specifically for webpack**

Using [this section](https://storybook.js.org/docs/writing-stories/build-pages-with-storybook#mocking-imports) of the docs
You can create a next `headers` mock (and by extension `cookies`)

**I've marked any new files with * their location doesn't matter**

```ts
// .storybook/withHeadersMock.ts *

import { Decorator } from '@storybook/react';

let allHeaders = new Map();
export const headers = () => {
  return allHeaders;
};

export const withNextHeadersMock: Decorator = (story, { parameters }) => {
  if (parameters?.nextHeaders) {
    allHeaders = new Map(Object.entries(parameters.nextHeaders));
  }

  return story();
};
```

```ts
// .storybook/main.ts
const config: StorybookConfig = {
  // rest of your config...
  webpackFinal: async (config) => {
    if (!config.resolve) {
      config.resolve = {};
    }

    if (!config.resolve.alias) {
      config.resolve.alias = {};
    }

    config.resolve.alias['next/headers'] = path.resolve(__dirname, './withNextHeadersMock.ts');
    return config;
  },
}
```

Then just add the headers you need to the story/preview
```ts
export const MyStory: Story = {
  decorators: [
    withNextHeadersMock
  ]
  parameters: {
    nextHeaders: {
      'my-header': 'foo'
    }
  }
};
```

This will at least give your stories the ability to run, albeit very mocked so only use this to tide over until a better solution is suggested / implemented

_Originally posted by @JamesManningR in https://github.com/storybookjs/storybook/issues/21540#issuecomment-1854045252_
            

### Are you able to assist to bring the feature to reality?

yes, I can

### Additional context

Mentioned here:
https://github.com/storybookjs/storybook/issues/21540#issuecomment-1854045252

**Comment by valentinpalkovic** [comment]
Is anyone interested in creating a PR to implement this?
